gobblin-service/src/main/java/org/apache/gobblin/service/modules/utils/ExecutionChecksUtil.java

```diff
@@ -0,0 +1,185 @@
+        eventSubmitter.transform(submitter -> new TimingEvent(submitter, TimingEvent.FlowTimings.FLOW_COMPILED));
+    Map<String, String> flowMetadata = TimingEventUtils.getFlowMetadata(flowSpec);
+
+    if (jobExecutionPlanDag == null || jobExecutionPlanDag.isEmpty()) {
+      populateFlowCompilationFailedEventMessage(eventSubmitter, flowSpec, flowMetadata);
+      return Optional.absent();
+    }
+
+    addFlowExecutionIdIfAbsent(flowMetadata, jobExecutionPlanDag);
+    if (flowCompilationTimer.isPresent()) {
+      flowCompilationTimer.get().stop(flowMetadata);
+    }
+    return Optional.of(jobExecutionPlanDag);
+  }
+
+  /**
+   * Checks if flowSpec disallows concurrent executions, and if so then checks if another instance of the flow is
+   * already running and emits a FLOW FAILED event. Otherwise, this check passes.
+   * @return true if caller can proceed to execute flow, false otherwise
+   * @throws IOException
+   */
+  public static boolean isExecutionPermittedHandler(SharedFlowMetricsContainer sharedFlowMetricsContainer,
+      SpecCompiler specCompiler, UserQuotaManager quotaManager, Optional<EventSubmitter> eventSubmitter,
+      FlowStatusGenerator flowStatusGenerator, Logger log, boolean flowConcurrencyFlag, Config flowConfig, Spec spec,
+      String flowName, String flowGroup) throws IOException {
+    boolean allowConcurrentExecution = ConfigUtils.getBoolean(flowConfig,
+        ConfigurationKeys.FLOW_ALLOW_CONCURRENT_EXECUTION, flowConcurrencyFlag);
+
+    Dag<JobExecutionPlan> jobExecutionPlanDag = specCompiler.compileFlow(spec);
+
+    if (!isExecutionPermitted(flowStatusGenerator, flowName, flowGroup, allowConcurrentExecution)) {
+      log.warn("Another instance of flowGroup: {}, flowName: {} running; Skipping flow execution since "
+          + "concurrent executions are disabled for this flow.", flowGroup, flowName);
+      sharedFlowMetricsContainer.conditionallyUpdateFlowGaugeSpecState(spec, SharedFlowMetricsContainer.CompiledState.SKIPPED);
+      Instrumented.markMeter(sharedFlowMetricsContainer.getSkippedFlowsMeter());
+      if (!((FlowSpec) spec).getConfigAsProperties().containsKey(ConfigurationKeys.JOB_SCHEDULE_KEY)) {
+        // For ad-hoc flow, we might already increase quota, we need to decrease here
+        for (Dag.DagNode dagNode : jobExecutionPlanDag.getStartNodes()) {
+          quotaManager.releaseQuota(dagNode);
+        }
+      }
+
+      // Send FLOW_FAILED event
+      Map<String, String> flowMetadata = TimingEventUtils.getFlowMetadata((FlowSpec) spec);
+      flowMetadata.put(TimingEvent.METADATA_MESSAGE, "Flow failed because another instance is running and concurrent "
+          + "executions are disabled. Set flow.allowConcurrentExecution to true in the flow spec to change this behaviour.");
+      if (eventSubmitter.isPresent()) {
+        new TimingEvent(eventSubmitter.get(), TimingEvent.FlowTimings.FLOW_FAILED).stop(flowMetadata);
+      }
+      return false;
+    }
+    return true;
```

Comment:
not sure where we are in the nesting, so that I've got this right... *but that's the problem I'm suggesting to solve here!* ;)

how about:
```
if (isExecPermitted(...)) {
  return true;
} else {
  ... // all the many other lines of code
}
```
?